File: Python/Exercice4_1.py
#coding: utf-8
import os
import sys

r, w = os.pipe()
fils = os.fork()

if fils == -1:
    print("Erreur dans fork")
    sys.exit(0)

if fils == 0:
    os.close(r)
    mot = input("Veuillez entrer un mot pour la recherche: ")
    os.write(w, mot.encode())
    os.close(w)
else:
    os.wait()
    os.close(w)
    bufferR = os.read(r, 256).decode().strip()
    os.close(r)

    nomfichier = "./text.txt"
    trouve = False
    try:
        with open(nomfichier, "r") as f:
            for ligne in f:
                for mot in ligne.split():
                    if mot == bufferR:
                        print(f'Le mot "{bufferR}" est dans le texte')
                        trouve = True
                        break
                if trouve:
                    break
        if not trouve:
            print(f'Le mot "{bufferR}" n\'est pas dans le texte')
    except FileNotFoundError:
        print("Fichier text.txt introuvable")

File: C/Exercice4_1.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>
#include <string.h>

#define BUFFER_SIZE 256

int main()
{
    FILE *file = NULL;
    __pid_t fils;
    int tube[2], i = 0;
    unsigned char bufferR[BUFFER_SIZE] = {0}, bufferW[BUFFER_SIZE] = {0};
    char mot[64], *nomfichier = "./text.txt";

    puts("Creation du tube\n");
    if(pipe(tube) != 0)
    {
        fprintf(stderr, "Erreur dans pipe\n");
        exit(0);
    }
    else
    {
        switch (fils = fork())
        {
            case -1:
                fprintf(stderr, "Erreur dans fork");
                exit(0);
                break;
            
            case 0:
                close(tube[0]);
                printf("Veuillez entrer un mot pour la recherche: ");
                scanf("%s", mot);
                sprintf(bufferW, "%s", mot);
                write(tube[1], bufferW, BUFFER_SIZE);

            default:
                wait(NULL);
                close(tube[1]);
                int n = read(tube[0], bufferR, BUFFER_SIZE);
                file = fopen(nomfichier, "r");
                while (fscanf(file, "%63s", mot) == 1)
                {
                    if(strcmp(mot, bufferR) == 0)
                    {
                        printf("Le mot \"%s\" est dans le texte\n", bufferR);
                        break;
                    }
                    else
                        i = 1;
                }
                if(i == 1)
                    printf("Le mot \"%s\" n'est pas dans le texte\n", bufferR);
                
                break;
        }
    }
    return 0;
}